WebView/app/src/androidTest/java/com/android/samples/webviewdemo/MainActivityTest.kt

```diff
@@ -0,0 +1,146 @@
+import androidx.test.espresso.web.webdriver.Locator
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import androidx.test.filters.MediumTest
+import androidx.test.rule.ActivityTestRule
+import kotlinx.coroutines.CompletableDeferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.runBlocking
+import org.hamcrest.CoreMatchers.`is`
+import org.hamcrest.CoreMatchers.containsString
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertTrue
+import org.junit.Rule
+import org.junit.Test
+import org.junit.runner.RunWith
+
+/**
+ * Launch, interact, and verify conditions in an activity that has a WebView instance.
+ */
+@RunWith(AndroidJUnit4::class)
+@MediumTest
+class MainActivityTest {
+
+    private val context = ApplicationProvider.getApplicationContext<Context>()
+
+    @Rule
+    @JvmField
+    val mainActivityRule = ActivityTestRule(MainActivity::class.java)
+
+    // Test to check that the drop down menu behaves as expected
+    @Test
+    fun dropDownMenu_SanFran() {
+        onWebView()
+            .forceJavascriptEnabled()
+            .withElement(findElement(Locator.ID, "location"))
+            .perform(webClick()) // Similar to perform(click())
+            .withElement(findElement(Locator.ID, "SF"))
+            .perform(webClick()) // Similar to perform(click())
+            .withElement(findElement(Locator.ID, "title"))
+            .check(webMatches(getText(), containsString("San Francisco")))
+    }
+
+    // Test for checking createJsObject
+    @Test
+    fun jsObjectIsInjectedAndContainsPostMessage() {
+        mainActivityRule.getActivity()
+        onWebView()
+            .check(
+                webMatches(
+                    script(
+                        "return jsObject && jsObject.postMessage != null;",
+                        castOrDie(Boolean::class.javaObjectType)
+                    ),
+                    `is`(true)
+                )
+            )
+    }
+
+    @Test
+    fun valueInCallback_compareValueInput_returnsTrue() = runBlocking {
+        // Setup
+        val jsObjName = "jsObject"
+        val allowedOriginRules = setOf("https://example.com")
+        val expectedMessage = "hello"
+        val onMessageReceived = CompletableDeferred<String?>()
+        launch(Dispatchers.Main) {
+            val webView = WebView(context).apply {
+                settings.javaScriptEnabled = true
+            }
+            // Create & inject JsObject
+            createJsObject(
+                webView,
+                jsObjName,
+                allowedOriginRules
+            ) { message ->
+                onMessageReceived.complete(message)
+            }
+            webView.loadDataWithBaseURL(
+                "https://example.com",
+                "<html><script>${jsObjName}.postMessage('${expectedMessage}')</script></html>",
+                "text/html",
+                "UTF-8",
```

Comment:
This isn't well documented at the moment, but please pass `null` for `encoding`.

From [loadData()](https://developer.android.com/reference/kotlin/android/webkit/WebView#loadData(kotlin.String,%20kotlin.String,%20kotlin.String)):

> _The encoding parameter specifies whether the data is base64 or URL encoded._ If the data is base64 encoded, the value of the encoding parameter must be "base64"... _For all other values of encoding (including `null`)_...

And from [loadDataWithBaseURL()](https://developer.android.com/reference/kotlin/android/webkit/WebView#loaddatawithbaseurl):

> If the base URL uses any other scheme, then the data will be loaded into the WebView as a plain string (i.e. not part of a data URL) and any URL-encoded entities in the string will not be decoded.